overlay/usr/lib/inithooks/bin/nextcloud.py

```diff
@@ -27,59 +30,158 @@ def usage(s=None):
     sys.exit(1)
 
 
+def set_password(password: str) -> tuple[int, str]:
+    """Set Nextcloud password - returns code & message
+    - if succcessful - return 0
+    - if password complexity failure - return 1
+    - if exception - return 2
+    - other unknown error - return 3
+    """
+    local_env = os.environ.copy()
+    local_env["OC_PASS"] = password
+    p = subprocess.run(
+        args=[
+            "/usr/local/bin/turnkey-occ",
+            "user:resetpassword",
+            "--password-from-env",
+            "admin",
+        ],
+        cwd="/var/www/nextcloud",
+        env=local_env,
+        text=True,
+        capture_output=True,
+    )
+
+    # password successfully set
+    if p.returncode == 0:
+        log.write("admin password set")
+        return (0, "")
+
+    # failure/error handling
+    else:
+        skip_pass = ("\nTo skip setting password: Skip-password"
+                     "\n- fix issues & set password manually")
+        rand_pass = ("\nTo set random password: Random1234"
+                     "\n- see inithooks log for password")
+        # occ seems to only output to stdout - even if stacktrace
+        # so write stdout to log regardless
+        log.write(p.stdout, "err")
+
+        # password complexity failure
+        if "password" in p.stdout.lower():
+            error_code = 1
+            msg = f"{p.stdout}{rand_pass}"
+
+        # nextcloud exception
+        elif "exception" in p.stdout.lower():
+            error_code = 2
+            head = "Nextcloud exception: "
+            tail = ("\nAre all services running?"
+                    f" See Inithooks log for info{skip_pass}")
+            if "redis" in p.stdout.lower():
+                tail = tail.format('redis')
+                msg = f"{head}Redis problem{tail}"
+            elif "database":
+                msg = f"{head}Database problem{tail}"
+            else:
+                msg = f"{head}Unknown problem{tail}"
+
+        # some other error
+        else:
+            error_code = 3
+            msg = f"Unexpected Nextcloud error{rand_pass}{skip_pass}"
+
+        return (error_code, f"\n{msg}")
```

Comment:
I think using integer literals here is less than ideal, using named literals might be better, by adding

```python
ERR_PASSWORD = 1
ERR_SERVICE = 2
ERR_UNKNOWN = 3

...

        if "password" in p.stdout.lower():
            error_code = ERR_PASSWORD
            msg = f"{p.stdout}{rand_pass}"
```

Rational is that by using named constants it becomes clearer to maintainers what these values mean at a glance. It also means that if a case changes meaning, is removed or a new one added, it will throw an unknown variable exception.

Reply:
Done :grin: